File: scala/P16.scala
// P16 Drop every N'th element from a list.

// Recursion
def f1[T](n: Int, list: List[T]): List[T] = {
  def f[T]: (Int, List[T]) => List[T] = {
    case (_, Nil) => Nil
    case (1, _ :: xs) => f(n, xs)
    case (k, x :: xs) => x :: f(k - 1, xs)
  }
  f(n, list)
}

// Tail recursion
def f2[T](n: Int, list: List[T]): List[T] = {
  @annotation.tailrec
  def f[T](k: Int, acc: List[T], l: List[T]): List[T] = (k, l) match {
    case (_, Nil) => acc.reverse
    case (1, _ :: xs) => f(n, acc, xs)
    case (k, x :: xs) => f(k - 1, x :: acc, xs)
  }
  f(n, Nil, list)
}

// Recursion with "take" and "drop"
def f3[T](n: Int, list: List[T]): List[T] =
  if (list.nonEmpty) list.take(n - 1) ::: f3(n, list.drop(n)) else Nil

// Zip with cycle, filter, map
def f4[T](n: Int, list: List[T]): List[T] = {
  val cycle = Stream.iterate(1)(it => if (it < n) it + 1 else 1)
  list.zip(cycle).filter(_._2 < n).map(_._1)
}

// Using "for comprehensions"
def f5[T](n: Int, list: List[T]): List[T] = {
  val cycle = Stream.iterate(1)(it => if (it < n) it + 1 else 1)
  for((x, k) <- list.zip(cycle) if k < n) yield x
}

// ZipWithIndex and modulo
def f6[T](n: Int, list: List[T]): List[T] =
  list.zipWithIndex.filter(it => (it._2 + 1) % n > 0).map(_._1)

// Imperative
def f7[T](n: Int, list: List[T]): List[T] = {
  var acc: List[T] = Nil
  var k = 1
  for (it <- list) {
    if (k < n) { 
      acc ::= it
      k += 1
    } else {
      k = 1
    }
  }
  acc.reverse
}

File: haskell/p16.hs
-- P16 Drop every N'th element from a list.

-- Recursion with a counter
f1 :: Int -> [a] -> [a]
f1 n = f n
    where f _ [] = []
          f 1 (_:xs) = f n xs
          f k (x:xs) = x : f (pred k) xs

-- Tail recursion
f2 :: Int -> [a] -> [a]
f2 n = f n []
    where f _ acc [] = reverse acc
          f 1 acc (_:xs) = f n acc xs
          f k acc (x:xs) = f (pred k) (x : acc) xs

-- Recursion with "take" and "drop"
f3 :: Int -> [a] -> [a]
f3 _ [] = []
f3 n xs = take (pred n) xs ++ (f2 n $ drop n xs)

-- Zip with cycle, filter, map
f4 :: Int -> [a] -> [a]
f4 n = map snd . filter ((< n) . fst) . zip (cycle [1..n])

-- Using list comprehensions
f5 :: Int -> [a] -> [a]
f5 n xs = [x | (x, k) <- zip xs $ cycle [1..n], k < n]